ckanext/canada/model.py

```diff
@@ -0,0 +1,87 @@
+# encoding: utf-8
+
+import datetime
+import logging
+
+from sqlalchemy import Column, Unicode, DateTime, Integer
+from sqlalchemy.ext.declarative import declarative_base
+
+from ckan.model import meta
+
+log = logging.getLogger(__name__)
+
+
+Base = declarative_base(metadata=meta.metadata)
+
+
+class PackageSync(Base):
+    __tablename__ = 'package_sync'
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    package_id = Column(Unicode)
+    created = Column(DateTime, default=datetime.datetime.now(datetime.timezone.utc))
+    last_successful_sync = Column(DateTime, nullable=True)
+    error_on = Column(Unicode, nullable=True)
+    error = Column(Unicode, nullable=True)
+
+    @classmethod
+    def get(cls, package_id, for_update=False):
+        '''Returns a package_sync object referenced by its package_id.'''
+        if not package_id:
+            return None
+
+        q = meta.Session.query(cls).autoflush(True).filter_by(package_id=package_id)
+        if for_update:
+            q = q.with_for_update()
+        return q.first()
+
+
+    @classmethod
+    def upsert(cls, package_id, last_successful_sync=None, error_on=None, error=None):
+        '''Sets and returns a package_sync object referenced by its package_id.'''
+        package_sync = cls.get(package_id, for_update=True)
+
+        if package_sync:
+            package_sync.error_on = error_on
+            package_sync.error = error
+            package_sync.created = datetime.datetime.now(datetime.timezone.utc)
+            package_sync.last_successful_sync = last_successful_sync
+        else:
+            package_sync = cls(package_id=package_id, last_successful_sync=last_successful_sync,
+                               error_on=error_on, error=error)
+
+        meta.Session.add(package_sync)
+        meta.Session.commit()
```

Comment:
I don't love having commit inside model methods because there's a side effect of committing all other changes that might be in the session, and looking at code that calls the method it's not obvious that that's happening.

Reply:
but, we're the only user of this so I suppose it's fine

Reply:
The CKAN Core DomainObject does this with the `save` method:

```
def save(self):
        self.add()
        self.commit()
```

kinda just based it off of that. But it does use `self.Session` / `cls.Session` which just saves a copy of `meta.Session` at class instantiation ?? I think?? Should I do it that way and have a `Session = meta.Session` and use `cls.Session` in the `PackageSync` model class?

Reply:
yeah it's just a pattern that I don't think is great because it doesn't mix well with the "dictization" code in ckan that adds things to the session without committing.

In this case it really doesn't matter, though might be slightly less efficient to commit after every row being created/updated.

Reply:
from testing, there are scenarios because we are in the CLI context in which we could exit the CLI and the commits would not take affect. @wardi I could handle the session commit in the `_quiet_int_pipe` context manager? that way if it does exit, then we commit to the DB?

Otherwise we can just keep with the commit per package inside the loop